src/borg/helpers/datastruct.py

```diff
@@ -49,3 +49,58 @@ def get(self, size=None, init=False):
         if size is not None:
             self.resize(size, init)
         return self.buffer
+
+
+class EfficientBytesQueue:
+    """
+    An efficient FIFO queue that splits received elements into chunks.
+    """
+
+    class SizeUnderflow(Error):
+        """Could not pop_front first {} elements, collection only has {} elements.."""
+
+    def __init__(self, split_size):
+        self.buffers = []
+        self.size = 0
+        self.split_size = split_size
+
+    def peek_front(self):
+        if not self.buffers:
+            return []
+        buffer = self.buffers[0]
+        return buffer
+
+    def pop_front(self, size):
+        if size > self.size:
+            raise EfficientBytesQueue.SizeUnderflow(size, self.size)
+        while size > 0:
+            buffer = self.buffers[0]
+            to_remove = min(size, len(buffer))
+            buffer = buffer[to_remove:]
+            if buffer:
+                self.buffers[0] = buffer
+            else:
+                self.buffers = self.buffers[1:]
+            size -= to_remove
+            self.size -= to_remove
+
+    def push_back(self, data):
+        if not self.buffers:
+            self.buffers = [b'']
+        while data:
+            buffer = self.buffers[-1]
+            if len(buffer) >= self.split_size:
+                buffer = b''
+                self.buffers = self.buffers + [buffer]
```

Comment:
is this needed? see next comment.

Reply:
If previous chunk is full, we have to make splitting point and grow "buffers". Otherwise we will always have only single buffer.
This is rationale for whole "efficiency". If we make single huge array then remove few elements from front, whole array has to be copied without first few elements. This is a problem if that array is hundreds of mebibytes (easily possible with `--remote-buffer 100`). So instead I create multiple arrays each at most 1mebibyte in length. Then if we have to make this copying, we never have to move more than 1mebibyte worth of data.

Reply:
ah, i was irritated by the `>= self.split_size` because it always will be `<=`, so only `==` will trigger this block.

but please keep it as is, it is more defensive like that.

the code in the block can be optimized to:
```
self.buffers.append(b'')  # start a new buffer when previous is full
```